statev2/state/src/applicator/order_book.rs

```diff
@@ -0,0 +1,488 @@
+
+    /// Reads the order info for the given order from storage
+    ///
+    /// Errors if the order is not present
+    fn read_order_info<T: TransactionKind>(
+        order_id: &OrderIdentifier,
+        tx: &DbTxn<'_, T>,
+    ) -> Result<NetworkOrder> {
+        let order_key = Self::order_key(order_id);
+        tx.read(ORDERS_TABLE, &order_key)
+            .map_err(StateApplicatorError::Storage)?
+            .ok_or_else(|| StateApplicatorError::MissingEntry(order_key))
+    }
+
+    /// Writes the order info for the given order to storage
+    fn write_order_info(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        let order_key = Self::order_key(&order.id);
+        tx.write(ORDERS_TABLE, &order_key, order)
+            .map_err(StateApplicatorError::Storage)
+    }
+
+    // --------------------------
+    // | Order Priority Helpers |
+    // --------------------------
+
+    /// Get the priority of a cluster
+    fn get_cluster_priority_with_tx<T: TransactionKind>(
+        cluster_id: &ClusterId,
+        tx: &DbTxn<'_, T>,
+    ) -> Result<u32> {
+        tx.read(PRIORITIES_TABLE, cluster_id)
+            .map_err(StateApplicatorError::Storage)
+            .map(|priority| priority.unwrap_or(CLUSTER_DEFAULT_PRIORITY))
+    }
+
+    /// Write an order priority to the DB
+    fn write_order_priority_with_tx(order: &NetworkOrder, tx: &DbTxn<'_, RW>) -> Result<()> {
+        // Lookup the cluster priority and write the order's priority
+        let cluster_priority = Self::get_cluster_priority_with_tx(&order.cluster, tx)?;
+        let priority = OrderPriority {
+            cluster_priority,
+            order_priority: ORDER_DEFAULT_PRIORITY,
+        };
+
+        tx.write(PRIORITIES_TABLE, &order.id, &priority)
+            .map_err(StateApplicatorError::Storage)
+    }
+
+    // -------------------------
+    // | Nullifier Set Helpers |
+    // -------------------------
+
+    /// Cancel all orders on a given nullifier
+    fn nullify_orders_with_tx(&self, nullifier: Scalar, tx: &DbTxn<'_, RW>) -> Result<()> {
+        let set = Self::read_nullifier_set(nullifier, tx)?;
+        for order_id in set.into_iter() {
+            self.cancel_order_with_tx(&order_id, tx)?;
+        }
+
+        Ok(())
+    }
+
+    /// Update the nullifier an order is indexed by
+    fn update_order_nullifier(
+        order_id: &OrderIdentifier,
+        old_nullifier: Scalar,
+        new_nullifier: Scalar,
+        tx: &DbTxn<'_, RW>,
+    ) -> Result<()> {
+        // Read the nullifier set for the old nullifier and remove the order
+        let old_set = Self::read_nullifier_set(old_nullifier, tx)?;
+        let updated_old_set = old_set
+            .into_iter()
+            .filter(|id| id != order_id)
+            .collect_vec();
+        Self::write_nullifier_set(old_nullifier, updated_old_set, tx)?;
+
+        // Add the order to the new nullifier set
+        Self::append_to_nullifier_set(new_nullifier, *order_id, tx)
```

Comment:
This feels unnecessarily DRY to me. It makes sense to use `append_to_nullifier_set` in `add_order_with_tx`, but using it here makes `update_order_nullifier` incur 2 reads and 2 writes where it only needs to incur one of each, and I don't think this helps readability all that much more than the alternative.

Reply:
It still needs 2 reads and 2 writes; and this is way nicer to work with than the raw deserialization and error handling that the method encapsulates